RationaleMCP/0036/README.md

```diff
@@ -39,19 +53,46 @@ One goal of "Clocked Discretized Continuous-Time Partition" is to preserve the b
 In addition the current reinit might possibly in the future be replaced by some form of impulses, indicating that the changes are physical and influence other parts of the system, this does not seem consistent with how states are set in this MCP.
 Finally having a conditional setting of states is more complicated, and reinit and other similar operators naturally suggest that - and we either need to determine what it means or forbid it.
```

Comment:
Please elaborate in what way conditional setting of states would be more complicated in this case compared to standard use of `reinit`.  To me, it seems that this could be handled very similar to `reinit`; the variable has dynamics that determine the value at the next tick, but can be reset to an observer's estimate at the end of any clock evaluation.

For the basic use, wouldn't it be sufficient to just allow eventless `reinit` of a clocked variable?  To make the picture complete, one would need to describe the allowed forms of clocked `reinit`.  The simplest thing to start with would be to only allow it in the form of an equation (a _clocked reinit-equation_), where there is no room for complicated things like conditionals.

Lastly, even if `reinit` isn't serving us well for what it is currently used for, perhaps it could still be the right concept for what this MCP is all about?

Reply:
> Please elaborate in what way conditional setting of states would be more complicated in this case compared to standard use of `reinit`. To me, it seems that this could be handled very similar to `reinit`; the variable has dynamics that determine the value at the next tick, but can be reset to an observer's estimate at the end of any clock evaluation.

Sure, I can elaborate - as I tried that first and realized the problems:
This MCP have simpler dynamics than above as it removed the dynamics for the variable completely (after index reduction), and only have the "reinit"-part.
And on the other hand, this "reinit" does not passively set a value for the state after everything else (as a normal reinit), but the update is part of a system of equations.

Reply:
It sounds as if the details of those examples would be really important for understanding what this MCP is really trying to achieve.  To start with, what is it that makes it necessary for the propagation of the estimate to be part of a system of equations?  It would be really interesting to see the detailed comparison between:
a. The current proposal (`x` and `xs` become aliases)
b. Propagating the estimate at the right side of a clock computation (one clock tick delay).
c. Propagating the estimate at the left side of a clock computation (no delay).

Can a real-world example motivate why the IMC should have alias-type access to the estimate rather than discrete-time propagation?

Reply:
After looking through the examples, and how the new value is used it became a bit clearer to me and my statement about "system of equations" was misleading.

Put simply: the inverse model contained `der(x)=f(x)`, and we simply want to update `x` and then compute `der(x)=f(x)` (which may involve a system of equations - but it's not important). (And `der(x)` is just an algebraic variable  - we don't integrate it.) So, the point is just that we want to use the new `x` already in the current step.

For a normal `reinit` we can currently rely on event iteration to compute a new `der(x)=f(x)` - but we don't do event iterations for clocked partitions. To have a description that we first integrate der(x)=f(x) (with a specified solver), and then ignore that, set x and compute der(x) might be possible, but I'm not sure it will work - and to me it seems that we are just complicating things (see note below).

Trying to unify this "set x first" and use it also for usual `reinit` seems really messy as the normal `reinit` is conditional and the conditions often depend on `x` which would cause a loop.

Note: One minor consequence of the current implementation is that if we have `der(y)=g(x, y)`; (I'm not sure if any of the examples contained that) we currently integrate that for the current step using the new value of `x`. This is clearly different from the idea of "first integrate, then update x (throwing away integration), then re-compute equations". I don't think this subtle change is important for most applications - but it indicates that "throwing away x" is not always fully possible and that such an alternative will sometimes be less efficient.

Note 2: It seems this estimate can also be applied in continuous time - and not only clocked. I guess that is more for testing control system with inverse models in continuous time than any real use.

Reply:
> Note: One minor consequence of the current implementation is that if we have `der(y)=g(x, y)`; (I'm not sure if any of the examples contained that) we currently integrate that for the current step using the new value of `x`.

I guess that makes sense for an implicit method, but would be more strange for an explicit method, right?

Still trying to convince myself that what is proposed on here isn't causing too much damage to the interpretation of the differential equations, I came up with the following — could this be an alternative characterization of what is currently implemented and described in the MCP?
- During a clocked integration step, the exact matching of `x` to `xs` is obtained by tweaking the discretized integration methods: instead of holding the previous value of `x` fixed, one holds the current value of `x` fixed (and equal to `xs`).  (The previous value of `x` becomes a dummy variable and should be discarded efter the computation.)

This way, one would be able to preserve the relation between `x` and `der(x)` during each integration step, and at the same time have `x = xs` after each step.